Multiple error types stored in atomic.Value

`atomic.Value` can't be used to store plain `error` interface values directly, because the dynamic type may differ between each call to Store.

```
panic: sync/atomic: store of inconsistently typed value into Value
2017/02/07 11:13:27 [DEBUG] plugin: terraform:
2017/02/07 11:13:27 [DEBUG] plugin: terraform: goroutine 59 [running]:
2017/02/07 11:13:27 [DEBUG] plugin: terraform: panic(0x2effba0, 0xc420330930)
2017/02/07 11:13:27 [DEBUG] plugin: terraform:  /usr/local/go/src/runtime/panic.go:500 +0x1a1
2017/02/07 11:13:27 [DEBUG] plugin: terraform: sync/atomic.(*Value).Store(0xc4203f0b80, 0x31686e0, 0xc420330920)
2017/02/07 11:13:27 [DEBUG] plugin: terraform:  /usr/local/go/src/sync/atomic/value.go:80 +0xd8
2017/02/07 11:13:27 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/provisioners/remote-exec.retryFunc.func1(0xc420402f60, 0x58fc320, 0xc420402f00, 0xc42040a5d0, 0xc4203f0b80)
2017/02/07 11:13:27 [DEBUG] plugin: terraform:  /Users/jake/src/go/src/github.com/hashicorp/terraform/builtin/provisioners/remote-exec/resource_provisioner.go:260 +0x162
2017/02/07 11:13:27 [DEBUG] plugin: terraform: created by github.com/hashicorp/terraform/builtin/provisioners/remote-exec.retryFunc
2017/02/07 11:13:27 [DEBUG] plugin: terraform:  /Users/jake/src/go/src/github.com/hashicorp/terraform/builtin/provisioners/remote-exec/resource_provisioner.go:262 +0x144
```

--- a/builtin/provisioners/remote-exec/resource_provisioner.go
+++ b/builtin/provisioners/remote-exec/resource_provisioner.go
@@ -234,10 +234,16 @@ func copyOutput(
 }
 
 // retryFunc is used to retry a function for a given duration
+// TODO: this should probably backoff too
 func retryFunc(ctx context.Context, timeout time.Duration, f func() error) error {
 	// Build a new context with the timeout
 	ctx, done := context.WithTimeout(ctx, timeout)
 	defer done()
+
+	// container for atomic error value
+	type errWrap struct {
+		E error
+	}
 
 	// Try the function in a goroutine
 	var errVal atomic.Value
@@ -255,19 +261,20 @@ func retryFunc(ctx context.Context, timeout time.Duration, f func() error) error
 
 			// Try the function call
 			err := f()
+			errVal.Store(&errWrap{err})
+
 			if err == nil {
 				return
 			}
 
 			log.Printf("Retryable error: %v", err)
-			errVal.Store(err)
 		}
 	}()
 
 	// Wait for completion
 	select {
-	case <-doneCh:
 	case <-ctx.Done():
+	case <-doneCh:
 	}
 
 	// Check if we have a context error to check if we're interrupted or timeout
@@ -279,8 +286,8 @@ func retryFunc(ctx context.Context, timeout time.Duration, f func() error) error
 	}
 
 	// Check if we got an error executing
-	if err, ok := errVal.Load().(error); ok {
-		return err
+	if ev, ok := errVal.Load().(errWrap); ok {
+		return ev.E
 	}
 
 	return nil

--- a/builtin/provisioners/remote-exec/resource_provisioner_test.go
+++ b/builtin/provisioners/remote-exec/resource_provisioner_test.go
@@ -2,8 +2,12 @@ package remoteexec
 
 import (
 	"bytes"
+	"context"
+	"errors"
 	"io"
+	"net"
 	"testing"
+	"time"
 
 	"github.com/hashicorp/terraform/config"
 	"github.com/hashicorp/terraform/helper/schema"
@@ -155,6 +159,31 @@ func TestResourceProvider_CollectScripts_scripts(t *testing.T) {
 		if out.String() != expectedScriptOut {
 			t.Fatalf("bad: %v", out.String())
 		}
+	}
+}
+
+func TestRetryFunc(t *testing.T) {
+	// succeed on the third try
+	errs := []error{io.EOF, &net.OpError{Err: errors.New("ERROR")}, nil}
+	count := 0
+
+	err := retryFunc(context.Background(), time.Second, func() error {
+		if count >= len(errs) {
+			return errors.New("failed to stop after nil error")
+		}
+
+		err := errs[count]
+		count++
+
+		return err
+	})
+
+	if count != 3 {
+		t.Fatal("retry func should have been called 3 times")
+	}
+
+	if err != nil {
+		t.Fatal(err)
 	}
 }
 

